segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -0,0 +1,635 @@
+ * Transactions are atomic - either all changes in the transaction are committed or none at all.
+ * In addition, Transactions provide snaphot isolation which means that transaction fails if any of the metadata records read during the transactions are changed outside the transaction after they were read.
+ *
+ * Within a transaction you can perform following actions on per record basis.
+ * <ul>
+ * <li>{@link MetadataTransaction#get(String)} Retrieves metadata using for given key.</li>
+ * <li>{@link MetadataTransaction#create(StorageMetadata)} Creates a new record.</li>
+ * <li>{@link MetadataTransaction#delete(String)} Deletes records for given key.</li>
+ * <li>{@link MetadataTransaction#update(StorageMetadata)} Updates the transaction local copy of the record.
+ * For each record modified inside the transaction update must be called to mark the record as dirty.</li>
+ * </ul>
+ * <div>
+ * <pre>
+ *  // Start a transaction.
+ * try (MetadataTransaction txn = metadataStore.beginTransaction()) {
+ *      // Retrieve the data from transaction
+ *      SegmentMetadata segmentMetadata = (SegmentMetadata) txn.get(streamSegmentName);
+ *
+ *      // Modify retrieved record
+ *      // seal if it is not already sealed.
+ *      segmentMetadata.setSealed(true);
+ *
+ *      // put it back transaction
+ *      txn.update(segmentMetadata);
+ *
+ *      // Commit
+ *      txn.commit();
+ *  } catch (StorageMetadataException ex) {
+ *      // Handle Exceptions
+ *  }
+ *  </pre>
+ * </div>
+ *
+ * Underlying implementation might buffer frequently or recently updated metadata keys to optimize read/write performance.
+ * To further optimize it may provide "lazy committing" of changes where there is application specific way to recover from failures.(Eg. when only length of chunk is changed.)
+ * In this case {@link MetadataTransaction#commit(boolean)} can be called.Note that otherwise for each commit the data is written to underlying key-value store.
+ *
+ * There are two special methods provided to handle metadata about data segments for the underlying key-value store. They are useful in avoiding circular references.
+ * <ul>
+ * <li>A record marked as pinned by calling {@link MetadataTransaction#markPinned(StorageMetadata)} is never written to underlying storage.</li>
+ * <li>In addition transaction can be committed using {@link MetadataTransaction#commit(boolean, boolean)} to skip validation step that reads any recently evicted changes from underlying storage.</li>
+ * </ul>
+ */
+@Slf4j
+abstract public class BaseMetadataStore implements ChunkMetadataStore {
+    /**
+     * Maximum number of metadata entries to keep in recent transaction buffer.
+     */
+    private static final int MAX_ENTRIES_IN_TXN_BUFFER = 5000;
+
+    /**
+     * List of pinned keys. These keys should not be read or written.
+     */
+    private final ConcurrentSkipListSet<String> pinnedKeys = new ConcurrentSkipListSet<>();
```

Comment:
Is this the most appropriate data structure?

Reply:
It is thread safe . It is log(n) and already provides floor() method. It it standard library.
It is implemented with less locking (lock free).  This data structure is used by lots of threads.
I think it's good enough for this purpose.